llvm/labs/lab2/vasilev_ivan/Vasilev.cpp

```diff
@@ -0,0 +1,90 @@
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+
+using namespace llvm;
+
+class MyLoopPass : public PassInfoMixin<MyLoopPass> {
+public:
+  PreservedAnalyses run(Function &Func, FunctionAnalysisManager &FAM) {
+    auto *Ty =
+        FunctionType::get(llvm::Type::getVoidTy(Func.getContext()), false);
+    auto *M = Func.getParent();
+
+    auto LoopStartFunc = M->getOrInsertFunction("_Z10loop_startv", Ty);
+    auto LoopEndFunc = M->getOrInsertFunction("_Z8loop_endv", Ty);
+
+    auto &LI = FAM.getResult<LoopAnalysis>(Func);
+
+    for (auto &Loop : LI) {
+      auto *Preheader = Loop->getLoopPreheader();
+
+      SmallVector<BasicBlock *, 4> ExitBlocks;
+      Loop->getExitBlocks(ExitBlocks);
+
+      if (!Preheader || ExitBlocks.empty())
+        continue;
+
+      int loop_start = 0;
+      int loop_end = 0;
+
+      for (Instruction &Instr : *Preheader) {
+        if (CallInst *CI = dyn_cast<CallInst>(&Instr)) {
+          if (CI->getCalledFunction() &&
```

Comment:
Please use searching in loop_start/loop_end funcs users lists for this check.

Reply:
I'm sorry, I'm not quite sure what you mean.